cpp/comms/std/src/cuML_std_comms_impl.cpp

```diff
@@ -0,0 +1,407 @@
+    case MLCommon::cumlCommunicator::INT:
+      return ncclInt;
+    case MLCommon::cumlCommunicator::UINT:
+      return ncclUint32;
+    case MLCommon::cumlCommunicator::INT64:
+      return ncclInt64;
+    case MLCommon::cumlCommunicator::UINT64:
+      return ncclUint64;
+    case MLCommon::cumlCommunicator::FLOAT:
+      return ncclFloat;
+    case MLCommon::cumlCommunicator::DOUBLE:
+      return ncclDouble;
+  }
+}
+
+ncclRedOp_t getNCCLOp(const cumlStdCommunicator_impl::op_t op) {
+  switch (op) {
+    case MLCommon::cumlCommunicator::SUM:
+      return ncclSum;
+    case MLCommon::cumlCommunicator::PROD:
+      return ncclProd;
+    case MLCommon::cumlCommunicator::MIN:
+      return ncclMin;
+    case MLCommon::cumlCommunicator::MAX:
+      return ncclMax;
+  }
+}
+}  // namespace
+
+bool ucx_enabled() { return UCX_ENABLED; }
+
+/**
+ * @brief Underlying comms, like NCCL and UCX, should be initialized and ready for use,
+ * and maintained, outside of the cuML Comms lifecycle. This allows us to decouple the
+ * ownership of the actual comms from cuml so that they can also be used outside of cuml.
+ *
+ * For instance, nccl-py can be used to bootstrap a ncclComm_t before it is
+ * used to construct a cuml comms instance. UCX endpoints can be bootstrapped
+ * in Python using ucx-py, before being used to construct a cuML comms instance.
+ */
+#ifdef WITH_UCX
+void inject_comms(cumlHandle &handle, ncclComm_t comm, ucp_worker_h ucp_worker,
+                  std::shared_ptr<ucp_ep_h *> eps, int size, int rank) {
+  auto communicator = std::make_shared<MLCommon::cumlCommunicator>(
+    std::unique_ptr<MLCommon::cumlCommunicator_iface>(
+      new cumlStdCommunicator_impl(comm, ucp_worker, eps, size, rank)));
+  handle.getImpl().setCommunicator(communicator);
+}
+#endif
+
+void inject_comms(cumlHandle &handle, ncclComm_t comm, int size, int rank) {
+  auto communicator = std::make_shared<MLCommon::cumlCommunicator>(
+    std::unique_ptr<MLCommon::cumlCommunicator_iface>(
+      new cumlStdCommunicator_impl(comm, size, rank)));
+  handle.getImpl().setCommunicator(communicator);
+}
+
+void inject_comms_py_coll(cumlHandle *handle, ncclComm_t comm, int size,
+                          int rank) {
+  inject_comms(*handle, comm, size, rank);
+}
+
+void inject_comms_py(ML::cumlHandle *handle, ncclComm_t comm, void *ucp_worker,
+                     void *eps, int size, int rank) {
+#ifdef WITH_UCX
+  std::shared_ptr<ucp_ep_h *> eps_sp =
+    std::make_shared<ucp_ep_h *>(new ucp_ep_h[size]);
+
+  size_t *size_t_ep_arr = (size_t *)eps;
+
+  for (int i = 0; i < size; i++) {
+    size_t ptr = size_t_ep_arr[i];
+    ucp_ep_h *ucp_ep_v = *eps_sp;
+
+    if (ptr != 0) {
+      ucp_ep_h *eps_ptr = (ucp_ep_h *)size_t_ep_arr[i];
+      ucp_ep_v[i] = *eps_ptr;
+    } else {
+      ucp_ep_v[i] = nullptr;
+    }
+  }
+
+  inject_comms(*handle, comm, (ucp_worker_h)ucp_worker, eps_sp, size, rank);
+
+#else
+  inject_comms(*handle, comm, size, rank);
```

Comment:
Does the else branch trigger compiler warnings as the arguments `ucp_worker` and `eps` are never used? To avoid this propably a `ifdef WITH_UCX` is also needed in the function header:
```cpp
void inject_comms_py(ML::cumlHandle *handle, ncclComm_t comm,
#ifdef WITH_UCX
                     void *ucp_worker, void *eps,
#else 
                     void *, void *,
#endif 
                     int size, int rank) {
```